### PR TITLE
Make HttpSM server reference a Transaction instead of a Session

### PR DESCRIPTION
**Pull request by shinrich**
This is derived as a subset of the H2 to origin PR #7622.  This PR adds no support for H2 to origin, but it performs the restructuring so the HttpSM refers to the server connection via a ProxyTransaction instead of a PoolableSession.  

This was requested by @maskit in hopes this work would assist them in discovering the cause of the server session leak described in issue #7471.  This restructuring should simplify server connection leaks because the server_txn object stays around until HttpSM::kill_this() when server_txn->transaction_done() is called.  If the server session should be returned to the origin pool, it will have previously been marked so.  Otherwise the server session will be closed.  Similarly earlier calls to do_io_close on the server transaction will be held until transaction_done() is called.  

In the case of an origin connection failure and subsequent retry, server_txn->transaction_done() is called in HttpSM::do_http_server_open() before another connection attempt is made.

The PR also includes some stats I added to help debug a pool buildup issue I ran into debugging this PR against our 9.1.x build in master.

I also pulled into the following commits to make the move to our 9.1.x branch from master easier.

PR #6241 - but I see that just got picked into 9.1.x already
PR #7571 - I think that one is there now too.
PR #7584 - which also just got pulled into 9.1.x
PR #7600

### CODE REVIEW

File: iocore/net/libinknet_stub.cc [hunk 1]
```diff
@@ -21,11 +21,11 @@
   limitations under the License.
  */
 
-#include "HttpSessionManager.h"
+class EThread;
+class Continuation;
```
Comment (shinrich) [inline iocore/net/libinknet_stub.cc:25]: Needed to avoid pulling in unnecessary stuff to the unit test.

File: proxy/PoolableSession.h [hunk 2]
```diff
@@ -192,3 +206,34 @@ PoolableSession::FQDNLinkage::equal(CryptoHash const &lhs, CryptoHash const &rhs
 {
   return lhs == rhs;
 }
+
```
Comment (shinrich) [inline proxy/PoolableSession.h:209]: Promoted the conntecting tracking calls to a super class.

File: proxy/http/Http1ServerSession.cc [hunk 3]
```diff
@@ -223,3 +189,71 @@ void
...
 }
+
+bool
+Http1ServerSession::is_chunked_encoding_supported() const
+{
+  return true;
+}
+
+void
+Http1ServerSession ::release_transaction()
```
Comment (shinrich) [inline proxy/http/Http1ServerSession.cc:200]: Called from transaction_done to adjust the released_transacations count and cause any pending close or server session pool release to occur.

File: proxy/http/HttpSM.cc [hunk 4]
```diff
@@ -3260,12 +3282,12 @@ HttpSM::is_bg_fill_necessary(HttpTunnelConsumer *c)
...
   if (c->producer->alive &&          // something there to read
                                      //      server_entry && server_entry->vc &&              // from an origin server
-                                     //      server_session && server_session->get_netvc() && // which is still open and valid
+                                     //      server_txn && server_txn->get_netvc() && // which is still open and valid
       c->producer->num_consumers > 1 // with someone else reading it
   ) {
     HttpTunnelProducer *p = nullptr;
 
-    if (!server_entry || !server_entry->vc || !server_session || !server_session->get_netvc()) {
+    if (!server_txn || !server_txn->get_netvc()) {
```
Comment (shinrich) [inline proxy/http/HttpSM.cc:3290]: May need to adjust this test, since server_txn will not be null at this point even if we did already call do_io_close on it.  The get_netvc() may still be returning null, so it may still be ok.

**Comment by shinrich** [comment]
[approve ci autest]

**Comment by shinrich** [comment]
[ci approve clang-analyzer]

**Comment by shinrich** [comment]
[approve ci clang-analyzer]

File: proxy/http/HttpSessionManager.cc [hunk 5]
```diff
@@ -484,3 +489,50 @@ HttpSessionManager::release_session(PoolableSession *to_release)
...
+
+void
+ServerSessionPool::removeSession(PoolableSession *to_remove)
+{
+  EThread *ethread = this_ethread();
+  SCOPED_MUTEX_LOCK(lock, mutex, ethread);
+  char peer_ip[INET6_ADDRPORTSTRLEN];
+  if (is_debug_tag_set("http_ss")) {
+    ats_ip_nptop(to_remove->get_remote_addr(), peer_ip, sizeof(peer_ip));
+    Debug("http_ss", "Remove session %p %s m_fqdn_pool size=%" PRId64 " m_ip_pool_size=%" PRId64, to_remove, peer_ip,
```
Comment (maskit) [inline proxy/http/HttpSessionManager.cc:501]: Please use `%zu` for `size_t`. https://github.com/apache/trafficserver/pull/7830

```
HttpSessionManager.cc:502:11: error: format specifies type 'long long' but the argument has type 'size_t' (aka 'unsigned long') [-Werror,-Wformat]
          m_fqdn_pool.count(), m_ip_pool.count());
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../include/tscore/Diags.h:158:35: note: expanded from macro 'Debug'
      log_message.debug(tag, loc, __VA_ARGS__);               \
                                  ^~~~~~~~~~~
HttpSessionManager.cc:502:32: error: format specifies type 'long long' but the argument has type 'size_t' (aka 'unsigned long') [-Werror,-Wformat]
          m_fqdn_pool.count(), m_ip_pool.count());
          ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
../../include/tscore/Diags.h:158:35: note: expanded from macro 'Debug'
      log_message.debug(tag, loc, __VA_ARGS__);               \
                                  ^~~~~~~~~~~
HttpSessionManager.cc:509:113: error: format specifies type 'long long' but the argument has type 'size_t' (aka 'unsigned long') [-Werror,-Wformat]
    Debug("http_ss", "After Remove session %p m_fqdn_pool size=%" PRId64 " m_ip_pool_size=%" PRId64, to_remove, m_fqdn_pool.count(),
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
../../include/tscore/Diags.h:158:35: note: expanded from macro 'Debug'
      log_message.debug(tag, loc, __VA_ARGS__);               \
                                  ^~~~~~~~~~~
HttpSessionManager.cc:510:11: error: format specifies type 'long long' but the argument has type 'size_t' (aka 'unsigned long') [-Werror,-Wformat]
          m_ip_pool.count());
          ^~~~~~~~~~~~~~~~~~
../../include/tscore/Diags.h:158:35: note: expanded from macro 'Debug'
      log_message.debug(tag, loc, __VA_ARGS__);               \
                                  ^~~~~~~~~~~
4 errors generated.
```

File: proxy/http/HttpSessionManager.cc [hunk 6]
```diff
@@ -484,3 +489,50 @@ HttpSessionManager::release_session(PoolableSession *to_release)
...
+    HTTP_DECREMENT_DYN_STAT(http_pooled_server_connections_stat);
+  }
+  if (is_debug_tag_set("http_ss")) {
+    Debug("http_ss", "After Remove session %p m_fqdn_pool size=%" PRId64 " m_ip_pool_size=%" PRId64, to_remove, m_fqdn_pool.count(),
+          m_ip_pool.count());
+  }
+}
+
+void
+ServerSessionPool::testSession(PoolableSession *ss)
```
Comment (maskit) [inline proxy/http/HttpSessionManager.cc:515]: It seems like this functions is unnecessary. It's called by `Http2ServerSession::test_session()` on #7622, but `Http2ServerSession::test_session()` is not called.

File: proxy/http/HttpSessionManager.h [hunk 7]
```diff
@@ -67,6 +67,9 @@ class ServerSessionPool : public Continuation
   static bool validate_host_sni(HttpSM *sm, NetVConnection *netvc);
   static bool validate_sni(HttpSM *sm, NetVConnection *netvc);
   static bool validate_cert(HttpSM *sm, NetVConnection *netvc);
+  void removeSession(PoolableSession *ssn);
```
Comment (maskit) [inline proxy/http/HttpSessionManager.h:70]: Can you make `removeSession` and `addSession` private functions for now because H1ServerSession doesn't need to call them directly?

I know the new functions are used by Http2ServerSession, and we can discuss the detail later on H2-to-Origin PR, but having `removeSession` and `addSession` as public functions in addition to `acquireSession` and `releaseSession` doesn't make sense to me. The interface would be confusing. I think the both H1ServerSession and H2ServerSession should be handled by HttpSessionManager consistently - unnecessary processes should be skipped inside HttpSessionManager.

File: proxy/ProxyTransaction.h [hunk 8]
```diff
@@ -85,6 +88,10 @@ class ProxyTransaction : public VConnection
   // Returns true if there is a request body for this request
   virtual bool has_request_body(int64_t content_length, bool is_chunked_set) const;
 
+  sockaddr const *get_remote_addr() const;
+
+  virtual HTTPVersion get_version(HTTPHdr &hdr) const;
```
Comment (maskit) [inline proxy/ProxyTransaction.h:93]: We don't need this at the moment.

Since`ProxySession::get_version` receives `HTTPHdr`, a proxy session can always return correct version. I wonder if there will be a session that has multiple versions of transactions. I'm not sure if we are allowed to use multiple versions on a traditional HTTP connection like below:

```
[Connect]
GET /a HTTP/1.0

GET /b HTTP/1.1
Host: example.com

[Close]
```

File: proxy/ProxyTransaction.h [hunk 9]
```diff
@@ -114,15 +121,17 @@ class ProxyTransaction : public VConnection
...
+  virtual void set_rx_error_code(ProxyError e);
+  virtual void set_tx_error_code(ProxyError e);
 
   bool support_sni() const;
 
   /// Variables
   //
   HttpSessionAccept::Options upstream_outbound_options; // overwritable copy of options
 
+  IOBufferReader *get_reader();
```
Comment (maskit) [inline proxy/ProxyTransaction.h:133]: It's almost bike-shedding and this is not the only place, but I don't like this name very much because the data source (or the direction) is unclear. My understanding is that this reader always reads data from a peer (i.e. a remote host) regardless of the type of transaction (ClientTransaction or ServerTransaction).

How about `get_remote_data_reader` ?

File: proxy/http/HttpSM.cc [hunk 10]
```diff
@@ -1800,6 +1798,65 @@ HttpSM::handle_api_return()
...
+  // session object.
+  if (s.outbound_conn_track_state.is_active()) {
+    Debug("http_connect", "[%" PRId64 "] max number of outbound connections: %d", this->sm_id, s.txn_conf->outbound_conntrack.max);
+    retval->enable_outbound_connection_tracking(s.outbound_conn_track_state.drop());
+  }
+  return retval;
+}
+
+void
+HttpSM::create_server_txn(NetVConnection *netvc, PoolableSession *new_session)
```
Comment (maskit) [inline proxy/http/HttpSM.cc:1841]: This function is really confusing.

The name implies that we call this function whenever we want to create a server transaction, but that is probably incorrect. This function seems to be only used for the initial server transaction on a session.

Furthermore, this is a public function despite that HttpSM has to be the only caller. Anybody can accidentally call this function based on the wrong assumption above.

Can you make this a private function and rename it to `create_initial_server_txn`? I haven't understood why server transactions have to be created here and there, and I think it should be done at only one place, but the name would tell that is the current design.

File: proxy/ProxyTransaction.cc [hunk 11]
```diff
@@ -221,6 +221,24 @@ ProxyTransaction::has_request_body(int64_t request_content_length, bool is_chunk
   return request_content_length > 0 || is_chunked;
 }
 
+bool
+ProxyTransaction::is_read_closed() const
```
Comment (zwoop) [inline proxy/ProxyTransaction.cc:225]: Is this used in some subsequent PRs? Seems a little odd to have this "no-op" method, which is never called as far as I can tell.

File: proxy/http/Http1ClientSession.cc [hunk 12]
```diff
@@ -345,11 +352,21 @@ Http1ClientSession::state_keep_alive(int event, void *data)
 {
   // Route the event.  It is either for vc or
   //  the origin server slave vc
-  if (data && data == slave_ka_vio) {
-    return state_slave_keep_alive(event, data);
-  } else {
-    ink_assert(data && data == ka_vio);
-    ink_assert(read_state == HCS_KEEP_ALIVE);
+  if (data) {
```
Comment (zwoop) [inline proxy/http/Http1ClientSession.cc:355]: Would this possibly read nicer as a switch statement?

File: proxy/http/HttpConfig.cc [hunk 13]
```diff
@@ -393,6 +393,25 @@ register_stat_callbacks()
...
+                     (int)http_origin_no_sharing, RecRawStatSyncCount);
+  RecRegisterRawStat(http_rsb, RECT_PROCESS, "proxy.process.http.origin.body", RECD_INT, RECP_NON_PERSISTENT, (int)http_origin_body,
+                     RecRawStatSyncCount);
+  RecRegisterRawStat(http_rsb, RECT_PROCESS, "proxy.process.http.origin.private", RECD_INT, RECP_NON_PERSISTENT,
+                     (int)http_origin_private, RecRawStatSyncCount);
+  RecRegisterRawStat(http_rsb, RECT_PROCESS, "proxy.process.http.origin.close_private", RECD_INT, RECP_NON_PERSISTENT,
+                     (int)http_origin_close_private, RecRawStatSyncCount);
+  RecRegisterRawStat(http_rsb, RECT_PROCESS, "proxy.process.http.origin.raw", RECD_INT, RECP_NON_PERSISTENT, (int)http_origin_raw,
+                     RecRawStatSyncCount);
+
```
Comment (zwoop) [inline proxy/http/HttpConfig.cc:414]: This is great!

File: proxy/http/HttpSM.cc [hunk 14]
```diff
@@ -1800,6 +1798,65 @@ HttpSM::handle_api_return()
   }
 }
 
+void
+HttpSM::set_server_txn(ProxyTransaction *txn)
+{
+  ink_release_assert(server_txn == nullptr);
```
Comment (zwoop) [inline proxy/http/HttpSM.cc:1804]: I would probably argue that this should be a debug assert, it kinda feels like we're too liberal sometimes with the release assert. Looking where this is called, it can't be nullptr (assuming that to_return->new_transaction() is not allowed to fail).

File: proxy/http/HttpSM.cc [hunk 15]
```diff
@@ -5095,11 +5123,26 @@ HttpSM::do_http_server_open(bool raw)
...
-      existing_ss->release(nullptr);
+      existing_ss->set_inactivity_timeout(HRTIME_SECONDS(t_state.txn_conf->keep_alive_no_activity_timeout_out));
+      existing_ss->release(server_txn);
       ua_txn->attach_server_session(nullptr);
     }
   }
+
+  if (!try_reuse) {
+    HTTP_INCREMENT_DYN_STAT(http_origin_make_new);
+    if (TS_SERVER_SESSION_SHARING_MATCH_NONE == t_state.txn_conf->server_session_sharing_match) {
```
Comment (zwoop) [inline proxy/http/HttpSM.cc:5134]: This code is slightly alarming; It feels like the server session would be in one of several conditions, as seen by which metric we increment. But, the logic to decide what "state" it is in is based on a number of conditions and member variables, rather than a single "state". That feels like e.g. someone could set "raw" and "is_private", and the behavior is now in a unknown state.

File: proxy/http/HttpSM.cc [hunk 16]
```diff
@@ -5492,14 +5534,13 @@ HttpSM::release_server_session(bool serve_from_cache)
...
   }
 
-  ink_assert(server_entry->vc == server_session);
-  server_entry->in_tunnel = true;
-  vc_table.cleanup_entry(server_entry);
-  server_entry   = nullptr;
-  server_session = nullptr;
+  if (server_entry) {
+    server_entry->vc       = nullptr;
+    server_entry->read_vio = server_entry->write_vio = nullptr;
```
Comment (zwoop) [inline proxy/http/HttpSM.cc:5541]: This would look a lot better if you don't do the double assignment on line 5541 :).

File: proxy/http/HttpSessionManager.cc [hunk 17]
```diff
@@ -165,9 +165,11 @@ ServerSessionPool::acquireSession(sockaddr const *addr, CryptoHash const &hostna
     }
     if (zret == HSM_DONE) {
       to_return = first;
-      HTTP_DECREMENT_DYN_STAT(http_pooled_server_connections_stat);
-      m_fqdn_pool.erase(first);
-      m_ip_pool.erase(to_return);
+      this->removeSession(to_return);
+    } else {
+      if (first != m_fqdn_pool.end()) {
```
Comment (zwoop) [inline proxy/http/HttpSessionManager.cc:170]: Bike shedding, but for easier read, why not make this an } else if { ?

File: proxy/http/Http1ServerSession.cc [hunk 18]
```diff
@@ -93,26 +98,19 @@ Http1ServerSession::new_connection(NetVConnection *new_vc, MIOBuffer *iobuf, IOB
...
-Http1ServerSession::enable_outbound_connection_tracking(OutboundConnTrack::Group *group)
+Http1ServerSession::do_io_close(int alerrno)
 {
-  ink_assert(nullptr == conn_track_group);
-  conn_track_group = group;
-  if (is_debug_tag_set("http_ss")) {
-    ts::LocalBufferWriter<256> w;
-    w.print("[{}] new connection, ip: {}, group ({}), count: {}\0", con_id, get_server_ip(), *group, group->_count);
-    Debug("http_ss", "%s", w.data());
+  if (state == SSN_CLOSED) { // Already been closed
```
Comment (maskit) [inline proxy/http/Http1ServerSession.cc:103]: I prefer this way. It has less dup code.
```
if (state != SSN_CLOSED) {
  state = SSN_CLOSED
  release outbound connection tracking
  close vc
  decrement stats
}

if (transact_count == released_transactions) {
  this->destroy();
}
```

File: proxy/http/HttpSessionManager.cc (hunk 6 above)
Comment (shinrich) [reply proxy/http/HttpSessionManager.cc:515]: Yes, this one can be removed.

File: proxy/http/HttpSessionManager.h (hunk 7 above)
Comment (shinrich) [reply proxy/http/HttpSessionManager.h:70]: Sure, in this use case they can be made private

File: proxy/ProxyTransaction.h (hunk 8 above)
Comment (shinrich) [reply proxy/ProxyTransaction.h:93]: I am not sure CONNECT makes my head hurt.  I would think that only one nested method would be interpreted over the CONNECT method.  And ATS would only interpret it if the CONNECT was made over a non-TLS connection.  Otherwise ATS treats the body as a blind tunnel.

Will have to read the RFC tea leaves on this.

File: proxy/ProxyTransaction.h (hunk 9 above)
Comment (shinrich) [reply proxy/ProxyTransaction.h:133]: I'm ok with get_remote_reader or remote_data_reader.  Just added this as a wrapper method around the _reader member.

File: proxy/http/HttpSM.cc (hunk 10 above)
Comment (shinrich) [reply proxy/http/HttpSM.cc:1841]: create_initial_server_txn is probably a better name.  And it can be made private.  The goal of the function is to wire in the netvc into a ProxyTransaction object (Http1ServerTransaction in this case).   It is taking logic that was inline in the HttpSM::state_http_server_open function.  It is called a bit later once H2 to origin is introduced because we need to wait to see what protocol is negotiated to create the correct subclass of ProxyTransaction.

File: proxy/ProxyTransaction.cc (hunk 11 above)
Comment (shinrich) [reply proxy/ProxyTransaction.cc:225]: Probably leftover from the H2 to origin pull back.

File: proxy/http/HttpSM.cc (hunk 14 above)
Comment (shinrich) [reply proxy/http/HttpSM.cc:1804]: Agree.  I use the release_asserts liberally while debugging in production.  Don't always get them cleaned up.  Looking forward to using your enable-all asserts option.

File: proxy/http/HttpSM.cc (hunk 15 above)
Comment (shinrich) [reply proxy/http/HttpSM.cc:5134]: I added these stats for debugging to solve my last issue before posting this PR.  No doubt could be tidied up.  Or removed entirely.

File: proxy/http/HttpSM.cc (hunk 10 above)
Comment (shinrich) [reply proxy/http/HttpSM.cc:1841]: Reworked so create_server_txn is called in HttpSessionManager too instead of calling new_transaction directly

File: proxy/ProxyTransaction.cc (hunk 11 above)
Comment (shinrich) [reply proxy/ProxyTransaction.cc:225]: Removed for now.  Will be added back with the H2 to origin logic.

File: proxy/http/Http1ClientSession.cc (hunk 12 above)
Comment (shinrich) [reply proxy/http/Http1ClientSession.cc:355]: Cannot have variables as case elements.

**Review by maskit — COMMENTED** [review]
Although I can't say this doesn't break things, but it looks good to me. I'm going to approve this when conflicts are resolved.

**Comment by shinrich** [comment]
[approve ci autest]

**Review by maskit — APPROVED** [review]
